pkg/relay/processmanager.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/flashbots/go-boost-utils/bls"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	ResponseQueueSubmit = iota
+	ResponseQueueRegister
+	ResponseQueueOther
+)
+
+type ProcessManager struct {
+	LastRegTime map[string]uint64 // [pubkey]timestamp
+	lrtl        sync.RWMutex      // LastRegTime RWLock
+
+	VerifySubmitBlockCh       chan VerifyReq
+	VerifyRegisterValidatorCh chan VerifyReq
+	VerifyOtherCh             chan VerifyReq
+
+	StoreCh chan StoreReq
+
+	m ProcessManagerMetrics
+}
+
+func NewProcessManager(verifySize, storeSize uint) *ProcessManager {
+	rm := &ProcessManager{
+		LastRegTime: make(map[string]uint64),
+
+		VerifySubmitBlockCh:       make(chan VerifyReq, verifySize),
+		VerifyRegisterValidatorCh: make(chan VerifyReq, verifySize),
+		VerifyOtherCh:             make(chan VerifyReq, verifySize),
+
+		StoreCh: make(chan StoreReq, storeSize),
+	}
+	rm.initMetrics()
+	return rm
+}
+
+func (rm *ProcessManager) RunVerify(num uint) {
+	for i := uint(0); i < num; i++ {
+		go rm.VerifyParallel()
+	}
+}
+
+func (rm *ProcessManager) RunStore(store Datastore, ttl time.Duration, num uint) {
+	for i := uint(0); i < num; i++ {
+		go rm.ParallelStore(store, ttl)
+	}
+}
+
+func (rm *ProcessManager) RunCleanup(checkinterval uint64, cleanupInterval time.Duration) {
+	for {
+		now := uint64(time.Now().Unix())
+		var keys []string
+		rm.lrtl.RLock()
+		for k, v := range rm.LastRegTime {
+			if checkinterval < now-v {
+				keys = append(keys, k)
+			}
+		}
+		rm.lrtl.RUnlock()
+
+		rm.lrtl.Lock()
+		for _, k := range keys {
+			delete(rm.LastRegTime, k)
+		}
+		rm.lrtl.Unlock()
+		rm.m.MapSize.Set(float64(len(rm.LastRegTime)))
+
+		time.Sleep(cleanupInterval)
+	}
+}
+
+func (rm *ProcessManager) LoadAll(m map[string]uint64) {
+	rm.lrtl.Lock()
+	defer rm.lrtl.Unlock()
+
+	for k, v := range m {
+		rm.LastRegTime[k] = v
+	}
+
+	rm.m.MapSize.Set(float64(len(rm.LastRegTime)))
+}
+
+func (rm *ProcessManager) GetStoreChan() chan StoreReq {
+	return rm.StoreCh
+}
+
+func (rm *ProcessManager) VerifyChan() chan VerifyReq {
+	return rm.VerifyOtherCh
+}
+
+func (rm *ProcessManager) GetVerifyChan(stack uint) chan VerifyReq {
+	switch stack {
+	case ResponseQueueSubmit:
+		return rm.VerifySubmitBlockCh
+	case ResponseQueueRegister:
+		return rm.VerifyRegisterValidatorCh
+	default: // ResponseQueueOther
+		return rm.VerifyOtherCh
+	}
+}
+
+func (rm *ProcessManager) Set(k string, value uint64) {
+	rm.lrtl.Lock()
+	defer rm.lrtl.Unlock()
+	defer rm.m.MapSize.Set(float64(len(rm.LastRegTime)))
+	rm.LastRegTime[k] = value
+}
+
+func (rm *ProcessManager) Get(k string) (value uint64, ok bool) {
+	rm.lrtl.RLock()
+	defer rm.lrtl.RUnlock()
+	value, ok = rm.LastRegTime[k]
+	return
+}
+
+func (rm *ProcessManager) ParallelStore(datas Datastore, ttl time.Duration) {
+	rm.m.RunningWorkers.WithLabelValues("ParallelStore").Inc()
+	defer rm.m.RunningWorkers.WithLabelValues("ParallelStore").Dec()
+	ctx := context.Background()
```

Comment:
Can we pass in a context instead of using `Background()` ?

Reply:
This context is local to the workers, and as there is no control on that so it's kind of done on purpose